lib/events/dynamoevents/dynamoevents.go

```diff
@@ -625,6 +781,233 @@ func (l *Log) getTableStatus(tableName string) (tableStatus, error) {
+		elapsed := time.Since(waitStart).Seconds()
+		log.Infof("Creating new DynamoDB index, %f seconds elapsed...", elapsed)
+	}
+
+	return nil
+}
+
+// removeV1GSI removes the pre RFD 24 global secondary index from the table.
+//
+// Invariants:
+// - The pre RFD 24 global secondary index must exist.
+// - This function must not be called concurrently with itself.
+// - This may only be executed after the post RFD 24 global secondary index has been created.
+func (l *Log) removeV1GSI() error {
+	c := dynamodb.UpdateTableInput{
+		TableName: aws.String(l.Tablename),
+		GlobalSecondaryIndexUpdates: []*dynamodb.GlobalSecondaryIndexUpdate{
+			{
+				Delete: &dynamodb.DeleteGlobalSecondaryIndexAction{
+					IndexName: aws.String(indexTimeSearch),
+				},
+			},
+		},
+	}
+
+	if _, err := l.svc.UpdateTable(&c); err != nil {
+		return trace.Wrap(convertError(err))
+	}
+
+	return nil
+}
+
+// migrateDateAttribute walks existing events and calculates the value of the new `date`
+// attribute and updates the event. This is needed by the new global secondary index
+// schema introduced in RFD 24.
+//
+// This function is not atomic on error but safely interruptible.
+// This means that the function may return an error without having processed
+// all data but no residual temporary or broken data is left and
+// the process can be resumed at any time by running this function again.
+//
+// Invariants:
+// - This function must be called after `createV2GSI` has completed successfully on the table.
+// - This function must not be called concurrently with itself.
+func (l *Log) migrateDateAttribute(ctx context.Context) error {
+	var total int64 = 0
+	var startKey map[string]*dynamodb.AttributeValue
+
+	for {
+		c := &dynamodb.ScanInput{
+			ExclusiveStartKey: startKey,
+			// Without consistent reads we may miss events as DynamoDB does not
+			// specify a sufficiently short synchronisation grace period we can rely on instead.
+			// This makes the scan operation slightly slower but the other alternative is scanning a second time
+			// for any missed events after an appropriate grace period which is far worse.
+			ConsistentRead: aws.Bool(true),
+			// 100 seems like a good batch size that compromises
+			// between memory usage and fetch frequency.
+			// The limiting factor in terms of speed is the update ratelimit and not this.
+			Limit:     aws.Int64(100),
+			TableName: aws.String(l.Tablename),
+			// Without the `date` attribute.
+			FilterExpression: aws.String("attribute_not_exists(CreatedAtDate)"),
+		}
+
+		// Resume the scan at the end of the previous one.
+		// This processes 100 events or 1 MiB of data at maximum
+		// which is why we need to run this multiple times on the dataset.
+		scanOut, err := l.svc.Scan(c)
+		if err != nil {
+			return trace.Wrap(convertError(err))
+		}
+
+		// For every item processed by this scan iteration we send an update action
+		// that adds the new date attribute.
+		for _, item := range scanOut.Items {
```

Comment:
Can this loop check the existing attributes on the event and skip events that were already migrated?

Reply:
The scan already filters out events that have the new attribute on the Dynamo side using the query `attribute_not_exists(CreatedAtDate)`. All events that this loop iterates over are not migrated.